testing/integration/upgrade_fleet_test.go

```diff
@@ -62,55 +71,135 @@ func TestFleetManagedUpgrade(t *testing.T) {
+			startVersionInfo.Binary.String(), startVersionInfo.Binary.Commit)
+	}
+
+	t.Logf("Testing Elastic Agent upgrade from %s to %s with Fleet...",
+		define.Version(), endVersionInfo.Binary.String())
+
+	testUpgradeFleetManagedElasticAgent(ctx, t, info, startFixture, endFixture, defaultPolicy())
+}
+
+func TestFleetAirGapedUpgrade(t *testing.T) {
+	stack := define.Require(t, define.Requirements{
+		Stack: &define.Stack{},
+		// The test uses iptables to simulate the air-gaped environment.
+		OS:      []define.OS{{Type: define.Linux}},
+		Isolate: true,  // Needed as the test blocks IPs using iptables.
+		Local:   false, // Needed as the test requires Agent installation
+		Sudo:    true,  // Needed as the test uses iptables and installs the Agent
+	})
+
+	ctx, _ := testcontext.WithDeadline(
+		t, context.Background(), time.Now().Add(10*time.Minute))
+
+	artifactAPI := tools.NewArtifactAPIClient()
+	latest, err := artifactAPI.GetLatestSnapshotVersion(ctx, t)
+	require.NoError(t, err, "could not fetch latest version from artifacts API")
+
+	// We need to prepare it first because it'll download the artifact, and it
+	// has to happen before we block the artifacts API IPs.
+	// The test does not need a fixture, but testUpgradeFleetManagedElasticAgent
+	// uses it to get some information about the agent version.
+	upgradeTo, err := atesting.NewFixture(
+		t,
+		latest.String(),
+		atesting.WithFetcher(atesting.ArtifactFetcher()),
+	)
+	require.NoError(t, err)
+	err = upgradeTo.Prepare(ctx)
+	require.NoError(t, err)
+
+	s := newArtifactsServer(ctx, t, latest.String())
+	host := "artifacts.elastic.co"
+	simulateAirGapedEnvironment(t, host)
+
+	rctx, cancel := context.WithTimeout(ctx, time.Second)
+	defer cancel()
+	req, err := http.NewRequestWithContext(rctx, http.MethodGet, "https://"+host, nil)
+	_, err = http.DefaultClient.Do(req)
+	if !(errors.Is(err, context.DeadlineExceeded) ||
+		errors.Is(err, os.ErrDeadlineExceeded)) {
+		t.Fatalf(
+			"request to %q should have failed, iptables rules should have blocked it",
+			host)
+	}
+
+	_, err = stack.ESClient.Info()
+	require.NoErrorf(t, err,
+		"failed to interact with ES after blocking %q through iptables", host)
+	_, body, err := stack.KibanaClient.Request(http.MethodGet, "/api/features",
+		nil, nil, nil)
+	require.NoErrorf(t, err,
+		"failed to interact with Kibana after blocking %q through iptables. "+
+			"It should not affect the connection to the stack. Host: %s, response body: %s",
+		stack.KibanaClient.URL, host, body)
+
+	fixture, err := define.NewFixture(t, define.Version())
+	require.NoError(t, err)
+	err = fixture.Prepare(ctx)
+	require.NoError(t, err)
+
+	t.Logf("Testing Elastic Agent upgrade from %s to %s with Fleet...",
+		define.Version(), latest)
+
+	downloadSource := kibana.DownloadSource{
+		Name:      "local-air-gaped-" + uuid.NewString(),
+		Host:      s.URL + "/downloads/beats/elastic-agent/",
```

Comment:
The route here should probably be a constant since it is also configured in the server.